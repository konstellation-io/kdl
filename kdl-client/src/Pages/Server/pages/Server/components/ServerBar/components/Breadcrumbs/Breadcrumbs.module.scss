@import 'Styles/variables';
@import 'Styles/mixins';
@import 'Styles/colors';

.container {
  display: flex;
  align-items: center;
  justify-content: space-between;
  & > div {
    margin-left: 2 * $grid-unit;
    &:not(:last-child):after {
      content: '/';
      margin-left: 2 * $grid-unit;
    }
  }
}
