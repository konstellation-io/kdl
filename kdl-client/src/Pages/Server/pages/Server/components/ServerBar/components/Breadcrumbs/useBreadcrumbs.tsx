import { useRouteMatch, useLocation } from 'react-router-dom';
import ROUTE from 'Constants/routes';
import { useQuery } from '@apollo/client';
import {
  GET_OPENED_SERVER,
  GetOpenedServer,
} from 'Graphql/client/queries/getOpenedServer.graphql';
import React from 'react';
import ServerIcon from 'Components/Icons/ServerIcon/ServerIcon';
import ServerIconStyle from 'Components/Icons/ServerIcon/ServerIcon.module.scss';
import ProjectIcon from 'Components/Icons/ProjectIcon/ProjectIcon';
import ProjectIconStyle from 'Components/Icons/ProjectIcon/ProjectIcon.module.scss';
import ExpandMoreIcon from '@material-ui/icons/ExpandMore';
import ServerMetrics from 'Pages/Server/pages/Server/components/ServerBar/components/ServerMetrics/ServerMetrics';
import { CrumbProps } from 'Pages/Server/pages/Server/components/ServerBar/components/Breadcrumbs/components/Crumb/Crumb';
import {
  GET_OPENED_PROJECT,
  GetOpenedProject,
} from 'Graphql/client/queries/getOpenedProject.graphql';
import useProjectNavigation, {
  EnhancedRouteConfiguration,
  projectRoutesConfiguration,
} from 'Hooks/useProjectNavigation';
import ProjectSelector from '../ProjectSelector/ProjectSelector';
import { GetProjects } from 'Graphql/queries/types/GetProjects';
import { loader } from 'graphql.macro';
import SectionSelector from '../SectionSelector/SectionSelector';
import cx from 'classnames';
const GetProjectsQuery = loader('Graphql/queries/getProjects.graphql');

function useBreadcrumbs() {
  const crumbs: CrumbProps[] = [];
  const routeMatch = useRouteMatch(ROUTE.PROJECT);
  const location = useLocation();
  const {
    data: projectsData,
    loading: projectsLoading,
    error: projectsError,
  } = useQuery<GetProjects>(GetProjectsQuery);
  const {
    data: projectData,
    loading: projectLoading,
    error: projectError,
  } = useQuery<GetOpenedProject>(GET_OPENED_PROJECT);
  const {
    data: serverData,
    loading: serverLoading,
    error: serverError,
  } = useQuery<GetOpenedServer>(GET_OPENED_SERVER);

  const {
    name: serverName,
    id: serverId,
    url: serverUrl,
    state: serverState,
  } = serverData?.openedServer || {
    name: '',
    id: '',
    url: '',
    state: '',
  };
  const openedProject = projectData?.openedProject || {
    id: '',
    name: '',
    state: '',
  };
  const projectSections: EnhancedRouteConfiguration[] = useProjectNavigation(
    serverId,
    openedProject.id
  );

  // Add server crumb
  crumbs.push({
    crumbText: serverName,
    LeftIconComponent: (
      <ServerIcon className={cx('icon-small', ServerIconStyle[serverState])} />
    ),
    BottomComponent: (
      <ServerMetrics serverUrl={serverUrl || ''} serverId={serverId} />
    ),
  });

  // Check if we are in a project
  if (routeMatch && openedProject) {
    // Add crumb for the project
    crumbs.push({
      crumbText: openedProject.name,
      LeftIconComponent: (
        <ProjectIcon
          className={cx('icon-small', ProjectIconStyle[openedProject.state])}
        />
      ),
      BottomComponent: (
        <ProjectSelector
          options={projectsData?.projects || []}
          selectedProjectId={openedProject.id}
          serverId={serverId}
        />
      ),
    });

    // Add crumb for the section
    const lastParam: string = location.pathname.split('/').pop() || '';
    const projectRoute = Object.values(projectRoutesConfiguration).find(
      ({ label }) => label.toLowerCase() === lastParam.toLowerCase()
    );

    if (projectRoute) {
      const { label: crumbText, Icon } = projectRoute;
      crumbs.push({
        crumbText,
        LeftIconComponent: <Icon className="icon-small" />,
        RightIconComponent: ExpandMoreIcon,
        BottomComponent: (
          <SectionSelector
            options={projectSections}
            selectedSection={lastParam}
          />
        ),
      });
    }
  }

  return {
    crumbs,
  };
}

export default useBreadcrumbs;
