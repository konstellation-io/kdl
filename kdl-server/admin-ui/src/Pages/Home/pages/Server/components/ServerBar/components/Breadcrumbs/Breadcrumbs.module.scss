@import 'Styles/variables';
@import 'Styles/mixins';
@import 'Styles/colors';

.container {
  position: relative;
  display: flex;
  align-items: center;
  justify-content: space-between;

  > div:not(:last-child):after {
    position: absolute;
    right: -6px;
    content: '/';
  }
}
