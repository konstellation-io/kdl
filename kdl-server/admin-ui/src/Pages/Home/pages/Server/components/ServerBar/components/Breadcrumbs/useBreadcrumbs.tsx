import {
  BottomComponentProps,
  CrumbProps,
} from 'Pages/Home/pages/Server/components/ServerBar/components/Breadcrumbs/components/Crumb/Crumb';
import {
  GET_OPENED_PROJECT,
  GetOpenedProject,
} from 'Graphql/client/queries/getOpenedProject.graphql';
import { useLocation, useRouteMatch } from 'react-router-dom';
import useProjectNavigation, {
  EnhancedRouteConfiguration,
  projectRoutesConfiguration,
} from 'Hooks/useProjectNavigation';
import ExpandMoreIcon from '@material-ui/icons/ExpandMore';
import { GetProjects } from 'Graphql/queries/types/GetProjects';
import ProjectIcon from 'Components/Icons/ProjectIcon/ProjectIcon';
import ProjectSelector from '../ProjectSelector/ProjectSelector';
import ROUTE from 'Constants/routes';
import React from 'react';
import SectionSelector from '../SectionSelector/SectionSelector';
import ServerIcon, {
  RemoteServerStates,
} from 'Components/Icons/ServerIcon/ServerIcon';
import ServerMetrics from 'Pages/Home/pages/Server/components/ServerBar/components/ServerMetrics/ServerMetrics';
import { loader } from 'graphql.macro';
import { useQuery } from '@apollo/client';
import { SERVER_NAME, SERVER_URL } from '../../../../../../../../index';

const GetProjectsQuery = loader('Graphql/queries/getProjects.graphql');

function useBreadcrumbs() {
  const crumbs: CrumbProps[] = [];
  const routeMatch = useRouteMatch(ROUTE.PROJECT);
  const location = useLocation();

  const {
    data: projectsData,
    loading: projectsLoading,
    error: projectsError,
  } = useQuery<GetProjects>(GetProjectsQuery);
  const {
    data: projectData,
    loading: projectLoading,
    error: projectError,
  } = useQuery<GetOpenedProject>(GET_OPENED_PROJECT);

  const projectSections: EnhancedRouteConfiguration[] = useProjectNavigation(
    '',
    projectData?.openedProject?.id || ''
  );

  const loading = projectLoading || projectsLoading;
  const error = projectError || projectsError;

  if (loading || !projectsData) return { loading, crumbs };
  if (error) throw Error('cannot retrieve data at useBreadcrumbs');

  // TODO: server env var
  // const {
  //   name: serverName,
  //   id: serverId,
  //   url: serverUrl,
  //   state: serverState,
  // } = serverData.openedServer;

  const openedProject = projectData?.openedProject;

  // Add server crumb
  crumbs.push({
    crumbText: SERVER_NAME,
    LeftIconComponent: (
      <ServerIcon
        className="icon-regular"
        state={RemoteServerStates.SIGNED_IN}
      />
    ),
    BottomComponent: (props: BottomComponentProps) => (
      <ServerMetrics serverUrl={SERVER_URL} {...props} />
    ),
  });

  // Check if we are in a project
  if (routeMatch && openedProject) {
    // Add crumb for the project
    const { name, state } = openedProject;
    crumbs.push({
      crumbText: name,
      LeftIconComponent: <ProjectIcon className="icon-regular" state={state} />,
      BottomComponent: (props: BottomComponentProps) => (
        <ProjectSelector options={projectsData.projects} {...props} />
      ),
    });

    // Add crumb for the section
    const lastParam: string = location.pathname.split('/').pop() || '';
    const projectRoute = Object.values(projectRoutesConfiguration).find(
      ({ id }) => id === lastParam
    );

    if (projectRoute) {
      const { label: crumbText, Icon } = projectRoute;
      crumbs.push({
        crumbText,
        LeftIconComponent: <Icon className="icon-small" />,
        RightIconComponent: ExpandMoreIcon,
        BottomComponent: (props: BottomComponentProps) => (
          <SectionSelector options={projectSections} {...props} />
        ),
      });
    }
  }

  return {
    crumbs,
  };
}

export default useBreadcrumbs;
