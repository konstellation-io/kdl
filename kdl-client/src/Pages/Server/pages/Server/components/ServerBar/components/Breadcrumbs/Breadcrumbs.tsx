import React from 'react';
import styles from './Breadcrumbs.module.scss';
import Crumb from './components/Crumb/Crumb';
import useBreadcrumbs from './useBreadcrumbs';

function Breadcrumbs() {
  const { crumbs } = useBreadcrumbs();
  return (
    <div className={styles.container}>
      {crumbs.map((crumb, index) => (
        <Crumb
          key={index}
          LeftIconComponent={crumb.LeftIconComponent}
          RightIconComponent={crumb.RightIconComponent}
          crumbText={crumb.crumbText}
          bottomComponent={crumb.bottomComponent}
        />
      ))}
    </div>
  );
}

export default Breadcrumbs;
