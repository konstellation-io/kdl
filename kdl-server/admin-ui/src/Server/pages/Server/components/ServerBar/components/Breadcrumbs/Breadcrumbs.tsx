import React from 'react';
import styles from './Breadcrumbs.module.scss';
import Crumb from './components/Crumb/Crumb';
import useBreadcrumbs from './useBreadcrumbs';

function Breadcrumbs() {
  const { crumbs } = useBreadcrumbs();
  console.log(crumbs);
  return (
    <div className={styles.container}>
      {crumbs.map((crumb, index) => (
        <Crumb {...crumb} key={index} />
      ))}
    </div>
  );
}

export default Breadcrumbs;
