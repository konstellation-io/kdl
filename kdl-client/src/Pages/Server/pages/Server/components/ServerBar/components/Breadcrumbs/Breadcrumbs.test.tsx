import Breadcrumbs from './Breadcrumbs';
import React from 'react';
import { shallow } from 'enzyme';

jest.mock('./useBreadcrumbs.tsx', () => () => ({
  crumbs: [
    {
      crumbText: 'foo',
      BottomComponent: <div>bar</div>,
      LeftIconComponent: <svg />,
    },
  ],
}));

let component;
beforeEach(() => {
  component = shallow(<Breadcrumbs />);
});

describe('Breadcrumbs component', () => {
  it('should render without crashing', function () {
    expect(component).toMatchSnapshot();
  });
});
