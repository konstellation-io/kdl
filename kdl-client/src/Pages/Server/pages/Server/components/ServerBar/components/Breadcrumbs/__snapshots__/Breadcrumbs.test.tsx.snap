// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Breadcrumbs component should render without crashing 1`] = `
<div
  className="container"
>
  <Crumb
    BottomComponent={
      <div>
        bar
      </div>
    }
    LeftIconComponent={<svg />}
    crumbText="foo"
    key="0"
  />
</div>
`;
